src/proc/mod.rs

```diff
@@ -588,29 +590,59 @@ impl GlobalCtx<'_> {
         handle: crate::Handle<crate::Expression>,
         arena: &crate::Arena<crate::Expression>,
     ) -> Result<u32, U32EvalError> {
+        match self.eval_expr_to_literal_from(handle, arena) {
+            Some(crate::Literal::U32(value)) => Ok(value),
+            Some(crate::Literal::I32(value)) => {
+                value.try_into().map_err(|_| U32EvalError::Negative)
+            }
+            _ => Err(U32EvalError::NonConst),
+        }
+    }
+
+    pub(crate) fn eval_expr_to_literal(
+        &self,
+        handle: crate::Handle<crate::Expression>,
+    ) -> Option<crate::Literal> {
+        self.eval_expr_to_literal_from(handle, self.const_expressions)
+    }
+
+    pub(crate) fn eval_expr_to_literal_from(
```

Comment:
What do you think of this?

```
    pub(crate) fn eval_expr_to_literal_from(
        &self,
        handle: crate::Handle<crate::Expression>,
        arena: &crate::Arena<crate::Expression>,
    ) -> Option<crate::Literal> {
        match arena[handle] {
            crate::Expression::Constant(c) => {
                self.eval_expr_to_literal_from(self.constants[c].init, self.const_expressions)
            }
            crate::Expression::Literal(literal) => Some(literal),
            crate::Expression::ZeroValue(ty) => match self.types[ty].inner {
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Sint,
                    width: 4,
                } => Some(crate::Literal::I32(0)),
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Uint,
                    width: 4,
                } => Some(crate::Literal::U32(0)),
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Float,
                    width: 4,
                } => Some(crate::Literal::F32(0.0)),
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Float,
                    width: 8,
                } => Some(crate::Literal::F64(0.0)),
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Bool,
                    width: 1,
                } => Some(crate::Literal::Bool(false)),
                _ => None,
            },
            _ => None,
        }
    }
```

Reply:
That is, just eliminate the internal function altogether in favor of recursion. The handle validation phase should have ensured that we have no loops.

Reply:
The family of functions this fn is part of are called by the frontends.
I find that the invariants are clearer without recursion.